src/main/kotlin/com/openlattice/shuttle/IntegrationService.kt

```diff
@@ -0,0 +1,279 @@
+private const val fetchSize = 10000
+private const val readRateLimit = 1000
+private const val uploadBatchSize = 10000
+private val nThreads = 2 * Runtime.getRuntime().availableProcessors()
+
+private lateinit var logEntityType: EntityType
+
+@Service
+class IntegrationService(
+        private val hazelcastInstance: HazelcastInstance,
+        private val missionParameters: MissionParameters,
+        private val idService: EntityKeyIdService,
+        private val entitySetManager: EntitySetManager,
+        private val reservationService: HazelcastAclKeyReservationService,
+        private val blackbox: Blackbox
+) {
+
+    private val integrations = hazelcastInstance.getMap<String, Integration>(HazelcastMap.INTEGRATIONS.name)
+    private val entitySets = hazelcastInstance.getMap<UUID, EntitySet>(HazelcastMap.ENTITY_SETS.name)
+    private val entityTypes = hazelcastInstance.getMap<UUID, EntityType>(HazelcastMap.ENTITY_TYPES.name)
+    private val propertyTypes = hazelcastInstance.getMap<UUID, PropertyType>(HazelcastMap.PROPERTY_TYPES.name)
+    private val integrationJobs = hazelcastInstance.getMap<UUID, IntegrationJob>(HazelcastMap.INTEGRATION_JOBS.name)
+    private val integrationQueue = hazelcastInstance.getQueue<UUID>(HazelcastQueue.INTEGRATION_JOBS.name)
+    private val creds = missionParameters.auth.credentials
+    private val semaphore = Semaphore(nThreads)
+    private val executor = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(nThreads))
+    private val statusPredicate = Predicates.or(
+            Predicates.equal(INTEGRATION_STATUS, IntegrationStatus.IN_PROGRESS),
+            Predicates.equal(INTEGRATION_STATUS, IntegrationStatus.QUEUED)
+    )
+
+    companion object {
+        @JvmStatic
+        fun init(blackbox: Blackbox, dataModelService: EdmManager) {
+            if (blackbox.enabled) {
+                logEntityType = dataModelService.getEntityType(FullQualifiedName(blackbox.entityTypeFqn))
+            }
+        }
+    }
+
+    init {
+        executor.execute {
+            semaphore.acquire()
+
+            //load any jobs that were in progress or queued
+            integrationJobs.keySet(statusPredicate).forEach{
+                integrationQueue.put(it)
+            }
+
+            while (true) {
+                if (semaphore.availablePermits() > 0) {
+                    loadCargo()
+                }
+            }
+        }
+    }
+
+    fun enqueueIntegrationJob(integrationName: String, integrationKey: UUID, callbackUrlAsString: Optional<String>): UUID {
+        checkState(integrations.containsKey(integrationName), "Integration with name $integrationName does not exist")
+        val integration = integrations.getValue(integrationName)
+        checkState(integrationKey == integration.key, "Integration key $integrationKey is incorrect")
+        var maybeCallbackUrl = Optional.empty<URL>()
+        callbackUrlAsString.ifPresent {
+            maybeCallbackUrl = Optional.of(URL(it))
```

Comment:
Should toss a try-catch around this b/c if `it` is a bad URL, you'll get an exception. 

Also I think it's worth considering doing this `URL` conversion here to ensure a valid `URL` but then storing the `URL` as a `String` in `IntegrationJob` since it's more compact than the structure of `URL` and it's only ever used (as a `URL`) at the end of the job anyway so you can just `URL()` it again at that point